BÀI TẬP CHƯƠNG 2: TRUY VẤN SQL NÂNG CAO

Họ và tên : Nguyễn Minh Hùng

MSV: 23174600148

Lớp: DHKL17A3HN

Cho hai bảng student(student_id, name, class, course_id, score) và course(id,
course_name)

yêu cầu: Sử dụng jupyter notebook để chạy code, hiển thị kết quả thu được và đưa ra kết luận/
câu trả lời về kết quả cuối cùng thu được.

kết nối SQLite và tạo bảng student và bảng course

In [2]:
import sqlite3
import pandas as pd
conn = sqlite3.connect("students.db")
cursor = conn.cursor()

In [3]:
cursor.execute("DROP TABLE IF EXISTS student;")
cursor.execute("DROP TABLE IF EXISTS course;")

# Tạo bảng student
cursor.execute("""
CREATE TABLE student (
    student_id INTEGER PRIMARY KEY,
    name TEXT,
    class TEXT,
    course_id INTEGER,
    score REAL
);
""")

# Tạo bảng course
cursor.execute("""
CREATE TABLE course (
    id INTEGER PRIMARY KEY,
    course_name TEXT
);
""")

# Lưu thay đổi vào database
conn.commit()

dữ liệu bảng student

In [4]:
student_data = [
    (1, "Nguyen Minh Hoang", "May Tinh", 12, 6.7),
    (2, "Tran Thi Lan", "Kinh Te", 34, 9.2),
    (3, "Pham Van Nam", "Toan Tin", None, 7.9),
    (4, "Le Thanh Huyen", "Toan Tin", 20, 7.2),
    (5, "Vu Quoc Anh", "May Tinh", 24, 8.0),
    (6, "Dang Thuy Linh", "May Tinh", 24, 5.5),
    (7, "Bui Tien Dung", "Kinh Te", 34, 9.2),
    (8, "Ho Ngoc Mai", "Toan Tin", 20, 8.8),
    (9, "Duong Huu Phuc", "Kinh Te", None, 7.2),
    (10, "Cao Thi Hanh", "May Tinh", None, 7.0)
]


dữ liệu bảng course

In [5]:
course_data = [
    (12, "Giai tich"),
    (34, "Thong ke"),
    (26, "Tin hoc")
]


chèn dữ liệu vào các bảng

In [6]:
course_data = [
    (12, "Giai tich"),
    (34, "Thong ke"),
    (26, "Tin hoc")
]


chèn dữ liệu vào 2 bảng

In [7]:
# Chèn dữ liệu vào bảng student
cursor.executemany("INSERT INTO student VALUES (?, ?, ?, ?, ?);", student_data)

# Chèn dữ liệu vào bảng course
cursor.executemany("INSERT INTO course VALUES (?, ?);", course_data)

conn.commit()

đọc dữ liệu

In [8]:
# Đọc dữ liệu từ bảng student
student_df = pd.read_sql_query("SELECT * FROM student;", conn)
print("Bảng Student:")
display(student_df)

# Đọc dữ liệu từ bảng course
course_df = pd.read_sql_query("SELECT * FROM course;", conn)
print("Bảng Course:")
display(course_df)

Bảng Student:


,student_id,name,class,course_id,score
0,1,Nguyen Minh Hoang,May Tinh,12.0,6.7
1,2,Tran Thi Lan,Kinh Te,34.0,9.2
2,3,Pham Van Nam,Toan Tin,NaN,7.9
3,4,Le Thanh Huyen,Toan Tin,20.0,7.2
4,5,Vu Quoc Anh,May Tinh,24.0,8.0
5,6,Dang Thuy Linh,May Tinh,24.0,5.5
6,7,Bui Tien Dung,Kinh Te,34.0,9.2
7,8,Ho Ngoc Mai,Toan Tin,20.0,8.8
8,9,Duong Huu Phuc,Kinh Te,NaN,7.2
9,10,Cao Thi Hanh,May Tinh,NaN,7.0


Bảng Course:


,id,course_name
0,12,Giai tich
1,26,Tin hoc
2,34,Thong ke


1. Bảng student

Cấu trúc:
student(student_id, name, class, course_id, score)

Ý nghĩa các cột:

student_id: mã sinh viên (khóa chính – định danh mỗi sinh viên).
name: tên sinh viên.
class: lớp hoặc ngành học (Máy Tính, Kinh Tế, Toán Tin).
course_id: mã môn học mà sinh viên đăng ký.
score: điểm của sinh viên ở môn học đó.

Nhận xét:

Bảng lưu thông tin sinh viên và điểm môn học.
course_id đóng vai trò khóa ngoại liên kết tới bảng course.
Một số bản ghi thiếu course_id (ví dụ sinh viên 3, 9, 10) → dữ liệu chưa đầy đủ.
Điểm số dao động khoảng 6.7 – 9.2, đa số sinh viên đạt trên 7.
Một số môn có nhiều sinh viên học:
course_id = 34 có nhiều sinh viên → môn này phổ biến.

2. Bảng course

Cấu trúc:
course(id, course_name)

Ý nghĩa các cột:

id: mã môn học (khóa chính).
course_name: tên môn học.

kết nối 2 bảng theo tích Decartes

In [9]:
print("Tích Decartes:")
cursor.execute('''
SELECT * FROM student, course;
''')
df_decartes = pd.DataFrame(cursor.fetchall(), columns=['student_id', 'name', 'class', 'course_id', 'score', 'id', 'course_name'])
print(df_decartes)
print("\n")


Tích Decartes:
    student_id               name     class  course_id  score  id course_name
0            1  Nguyen Minh Hoang  May Tinh       12.0    6.7  12   Giai tich
1            1  Nguyen Minh Hoang  May Tinh       12.0    6.7  26     Tin hoc
2            1  Nguyen Minh Hoang  May Tinh       12.0    6.7  34    Thong ke
3            2       Tran Thi Lan   Kinh Te       34.0    9.2  12   Giai tich
4            2       Tran Thi Lan   Kinh Te       34.0    9.2  26     Tin hoc
5            2       Tran Thi Lan   Kinh Te       34.0    9.2  34    Thong ke
6            3       Pham Van Nam  Toan Tin        NaN    7.9  12   Giai tich
7            3       Pham Van Nam  Toan Tin        NaN    7.9  26     Tin hoc
8            3       Pham Van Nam  Toan Tin        NaN    7.9  34    Thong ke
9            4     Le Thanh Huyen  Toan Tin       20.0    7.2  12   Giai tich
10           4     Le Thanh Huyen  Toan Tin       20.0    7.2  26     Tin hoc
11           4     Le Thanh Huyen  Toan Tin      

Kết quả hiển thị 30 dòng là đúng với phép tích Descartes.
Dữ liệu bị lặp lại nhiều lần vì mỗi sinh viên được ghép với mọi môn học.
Phép này không thể hiện đúng quan hệ thực tế, nên thường cần thêm điều kiện JOIN để liên kết chính xác giữa student.course_id và course.id.

kết nối 2 bảng sử dụng JOIN: INNER JOIN, LEFT JOIN, RIGHT JOIN, FULL OUTER JOIN

* INNER JOIN

In [10]:
print("INNER JOIN:")
cursor.execute('''
SELECT * FROM student INNER JOIN course ON student.course_id = course.id;
''')
df_inner = pd.DataFrame(cursor.fetchall(), columns=['student_id', 'name', 'class', 'course_id', 'score', 'id', 'course_name'])
print(df_inner)
print("\n")

INNER JOIN:
   student_id               name     class  course_id  score  id course_name
0           1  Nguyen Minh Hoang  May Tinh         12    6.7  12   Giai tich
1           2       Tran Thi Lan   Kinh Te         34    9.2  34    Thong ke
2           7      Bui Tien Dung   Kinh Te         34    9.2  34    Thong ke




Chỉ những bản ghi có mã môn học trùng khớp giữa hai bảng mới được hiển thị.
Trong dữ liệu hiện tại chỉ có 3 sinh viên có course_id trùng với id trong bảng course, nên kết quả trả về 3 dòng.

* LEFT JOIN

In [11]:
print("LEFT JOIN:")
cursor.execute('''
SELECT * FROM student LEFT JOIN course ON student.course_id = course.id;
''')
df_left = pd.DataFrame(cursor.fetchall(), columns=['student_id', 'name', 'class', 'course_id', 'score', 'id', 'course_name'])
print(df_left)
print("\n")

LEFT JOIN:
   student_id               name     class  course_id  score    id course_name
0           1  Nguyen Minh Hoang  May Tinh       12.0    6.7  12.0   Giai tich
1           2       Tran Thi Lan   Kinh Te       34.0    9.2  34.0    Thong ke
2           3       Pham Van Nam  Toan Tin        NaN    7.9   NaN        None
3           4     Le Thanh Huyen  Toan Tin       20.0    7.2   NaN        None
4           5        Vu Quoc Anh  May Tinh       24.0    8.0   NaN        None
5           6     Dang Thuy Linh  May Tinh       24.0    5.5   NaN        None
6           7      Bui Tien Dung   Kinh Te       34.0    9.2  34.0    Thong ke
7           8        Ho Ngoc Mai  Toan Tin       20.0    8.8   NaN        None
8           9     Duong Huu Phuc   Kinh Te        NaN    7.2   NaN        None
9          10       Cao Thi Hanh  May Tinh        NaN    7.0   NaN        None




LEFT JOIN giữ toàn bộ dữ liệu của bảng bên trái (student).
Chỉ những bản ghi có mã môn học phù hợp mới hiển thị thông tin môn học từ bảng course.
Các bản ghi không khớp vẫn được giữ lại nhưng giá trị môn học sẽ là NULL/None.

* RIGHT JOIN

In [12]:
print("RIGHT JOIN:")
cursor.execute('''
SELECT * FROM course LEFT JOIN student ON course.id = student.course_id;
''')
df_right = pd.DataFrame(cursor.fetchall(), columns=['id', 'course_name', 'student_id', 'name', 'class', 'course_id', 'score'])
print(df_right)
print("\n")

RIGHT JOIN:
   id course_name  student_id               name     class  course_id  score
0  12   Giai tich         1.0  Nguyen Minh Hoang  May Tinh       12.0    6.7
1  26     Tin hoc         NaN               None      None        NaN    NaN
2  34    Thong ke         7.0      Bui Tien Dung   Kinh Te       34.0    9.2
3  34    Thong ke         2.0       Tran Thi Lan   Kinh Te       34.0    9.2




* FULL OUTER JOIN

In [13]:
query = '''
SELECT * FROM student LEFT JOIN course
ON student.course_id = course.id
UNION ALL
SELECT * FROM course LEFT JOIN student
ON course.id = student.course_id
WHERE student.course_id IS NULL;
'''
df = pd.read_sql_query(query, conn)
print("FULL OUTER JOIN: ", df)

FULL OUTER JOIN:      student_id               name     class  course_id  score    id  \
0            1  Nguyen Minh Hoang  May Tinh       12.0    6.7  12.0   
1            2       Tran Thi Lan   Kinh Te       34.0    9.2  34.0   
2            3       Pham Van Nam  Toan Tin        NaN    7.9   NaN   
3            4     Le Thanh Huyen  Toan Tin       20.0    7.2   NaN   
4            5        Vu Quoc Anh  May Tinh       24.0    8.0   NaN   
5            6     Dang Thuy Linh  May Tinh       24.0    5.5   NaN   
6            7      Bui Tien Dung   Kinh Te       34.0    9.2  34.0   
7            8        Ho Ngoc Mai  Toan Tin       20.0    8.8   NaN   
8            9     Duong Huu Phuc   Kinh Te        NaN    7.2   NaN   
9           10       Cao Thi Hanh  May Tinh        NaN    7.0   NaN   
10          26            Tin hoc      None        NaN    NaN   NaN   

   course_name  
0    Giai tich  
1     Thong ke  
2         None  
3         None  
4         None  
5         None  
6     Thon

FULL OUTER JOIN giữ toàn bộ dữ liệu của cả hai bảng.
Các bản ghi khớp nhau sẽ được kết hợp lại.
Các bản ghi không khớp vẫn xuất hiện nhưng các cột còn lại sẽ có giá trị NULL/NaN.
Phép JOIN này giúp xem đầy đủ tất cả sinh viên và tất cả môn học, kể cả những dữ liệu chưa liên kết với nhau.

 bài 2. Hãy cập nhật những giá trị course_id còn thiếu trong bảng student bằng câu lệnh SQL,
trong đó các giá trị được điền là những giá trị nằm trong bảng course và loại bỏ những
bản ghi tham gia những môn học không tồn tại bảng course.

In [14]:
# Cập nhật course_id còn thiếu bằng giá trị ngẫu nhiên từ bảng course
cursor.execute('''
UPDATE student
SET course_id = (
    SELECT id FROM course ORDER BY RANDOM() LIMIT 1
)
WHERE course_id IS NULL;
''')
conn.commit()

# Loại bỏ các bản ghi có course_id không hợp lệ
cursor.execute('''
DELETE FROM student
WHERE course_id NOT IN (SELECT id FROM course);
''')
conn.commit()

# Kiểm tra lại dữ liệu
cursor.execute('SELECT * FROM student;')
print(pd.DataFrame(cursor.fetchall(), columns=['student_id', 'name', 'class', 'course_id', 'score']))

   student_id               name     class  course_id  score
0           1  Nguyen Minh Hoang  May Tinh         12    6.7
1           2       Tran Thi Lan   Kinh Te         34    9.2
2           3       Pham Van Nam  Toan Tin         26    7.9
3           7      Bui Tien Dung   Kinh Te         34    9.2
4           9     Duong Huu Phuc   Kinh Te         26    7.2
5          10       Cao Thi Hanh  May Tinh         26    7.0


Bài toán yêu cầu cập nhật các giá trị course_id còn thiếu trong bảng student bằng một giá trị ngẫu nhiên từ bảng course, sau đó loại bỏ các bản ghi có course_id không tồn tại trong bảng course.

Kết quả thực hiện cho thấy các bản ghi thiếu course_id đã được cập nhật thành công, đồng thời các sinh viên đăng ký những môn học không hợp lệ đã được loại bỏ. Sau khi xử lý, bảng student chỉ còn các bản ghi có course_id hợp lệ, đảm bảo tính nhất quán và toàn vẹn dữ liệu giữa hai bảng.

a, tổng số sinh viên điểm trung bình từng lớp

In [15]:
query = '''
SELECT class, COUNT(student_id) AS "tổng sinh viên", ROUND(AVG(score), 2) AS "điểm trung bình"
FROM student
GROUP BY class;
'''

df = pd.read_sql_query(query, conn)
print(df)

      class  tổng sinh viên  điểm trung bình
0   Kinh Te               3             8.53
1  May Tinh               2             6.85
2  Toan Tin               1             7.90


Câu truy vấn đã nhóm dữ liệu theo từng lớp để tính tổng số sinh viên và điểm trung bình của mỗi lớp. Kết quả cho thấy lớp Kinh Tế có số sinh viên nhiều nhất và cũng có điểm trung bình cao nhất. Lớp Máy Tính có điểm trung bình thấp hơn, trong khi lớp Toán Tin chỉ có một sinh viên nên điểm trung bình chính là điểm của sinh viên đó.

b, tổng số sinh viên, điểm trung bình từng môn học

In [16]:
query='''
SELECT course.course_name,
COUNT(student.student_id) AS "tổng sinh viên",
ROUND(AVG(student.score),2) AS "điểm trung bình môn học"
FROM course
LEFT JOIN student ON student.course_id = course.id
GROUP BY course.course_name;
'''
df = pd.read_sql_query(query, conn)
print(df)

  course_name  tổng sinh viên  điểm trung bình môn học
0   Giai tich               1                     6.70
1    Thong ke               2                     9.20
2     Tin hoc               3                     7.37


Giải tích

-Có 1 sinh viên tham gia.
Điểm trung bình 6.7.

Thống kê

-Có 2 sinh viên tham gia.
-Điểm trung bình 9.2.
-Đây là môn có nhiều sinh viên học nhất và điểm trung bình cao nhất.

Tin học

-Có 3 sinh viên tham gia
điểm trung bình 7.37

c, phân loại thi đua theo số điểm
- i. Điểm TB ≥  9.0: Xuất sắc.
- ii. 6.0 ≤ Điểm TB ≤ 8.9: Tốt.
- iii. Điểm TB < 6.0: Kém.

In [17]:
query = '''
SELECT course.course_name,
ROUND(AVG(student.score),2) AS "điểm trung bình",
CASE
    WHEN ROUND(AVG(student.score),2) >= 9.0 THEN "Xuất sắc"
    WHEN ROUND(AVG(student.score),2) BETWEEN 6.0 AND 8.9 THEN "Tốt"
    ELSE "Kém"
END AS "điểm"
FROM course
LEFT JOIN student ON student.course_id = course.id
GROUP BY course.course_name;
'''

df = pd.read_sql_query(query, conn)
print(df)


  course_name  điểm trung bình      điểm
0   Giai tich             6.70       Tốt
1    Thong ke             9.20  Xuất sắc
2     Tin hoc             7.37       Tốt


-Kết quả cho thấy môn Giải tích có điểm trung bình 6.7 và được xếp loại Tốt, môn Tin học có điểm trung bình 7.37 đạt loại Tốt . Điều này cho thấy sinh viên có kết quả học tập khá tốt ở hai môn này.
- đặc biệt, môn Thống kê đạt điêm trung bình 9.2 đạt loại Xuất sắc cho thấy sinh viên học môn này ở mức rất tốt.


BÀI 3 : Xếp hạng sinh viên

a, Điểm số

In [18]:
# Truy vấn xếp hạng sinh viên theo điểm số
query = '''
SELECT student_id, name, class, course_id, score,
       RANK() OVER (ORDER BY score DESC) AS rank
FROM student
ORDER BY score DESC;
'''

df_rank = pd.read_sql_query(query, conn)
print(df_rank)


   student_id               name     class  course_id  score  rank
0           2       Tran Thi Lan   Kinh Te         34    9.2     1
1           7      Bui Tien Dung   Kinh Te         34    9.2     1
2           3       Pham Van Nam  Toan Tin         26    7.9     3
3           9     Duong Huu Phuc   Kinh Te         26    7.2     4
4          10       Cao Thi Hanh  May Tinh         26    7.0     5
5           1  Nguyen Minh Hoang  May Tinh         12    6.7     6


Điểm số của sinh viên dao động trong khoảng 6.7 đến 9.2.

Trần Thị Lan và Bùi Tiến Dũng đạt điểm cao nhất (9.2) nên cùng đứng hạng 1.
Phạm Văn Nam đạt 7.9, đứng hạng 3, thuộc nhóm điểm khá.

Dương Hữu Phúc và Cao Thị Hạnh có điểm lần lượt 7.2 và 7.0, nằm ở mức trung bình khá.

Nguyễn Minh Hoàng có điểm thấp nhất (6.7) và đứng hạng 6.

Kết luận chung:

Kết quả cho thấy phần lớn sinh viên có mức điểm khá trở lên, không có điểm quá thấp. Đồng thời có hiện tượng đồng hạng ở vị trí cao nhất do hai sinh viên đạt cùng điểm số. Điều này phản ánh sự chênh lệch điểm giữa các sinh viên không quá lớn.

b, Điểm số theo lớp học

In [19]:
query = '''
SELECT student_id, name, class, course_id, score,
       RANK() OVER (PARTITION BY class ORDER BY score DESC) AS class_rank
FROM student
ORDER BY class, class_rank;
'''

df_class_rank = pd.read_sql_query(query, conn)
print(df_class_rank)


   student_id               name     class  course_id  score  class_rank
0           2       Tran Thi Lan   Kinh Te         34    9.2           1
1           7      Bui Tien Dung   Kinh Te         34    9.2           1
2           9     Duong Huu Phuc   Kinh Te         26    7.2           3
3          10       Cao Thi Hanh  May Tinh         26    7.0           1
4           1  Nguyen Minh Hoang  May Tinh         12    6.7           2
5           3       Pham Van Nam  Toan Tin         26    7.9           1


Dựa vào bảng dữ liệu, sinh viên được xếp hạng điểm số trong từng lớp

Lớp Kinh Tế
Trần Thị Lan và Bùi Tiến Dũng có điểm 9.2, cùng đứng hạng 1 trong lớp.

Dương Hữu Phúc có điểm 7.2, đứng hạng 3 do có hai sinh viên cùng điểm cao hơn.

Lớp Máy Tính

Cao Thị Hạnh có điểm 7.0, đứng hạng 1 trong lớp.

Nguyễn Minh Hoàng có điểm 6.7, đứng hạng 2.

Lớp Toán Tin

Phạm Văn Nam có điểm 7.9, là sinh viên duy nhất trong lớp, nên đứng hạng 1.

Kết luận chung

Bảng kết quả theo từng lớp đã xếp hạng chính xác sinh viên dựa trên điểm số trong phạm vi từng lớp. Một số lớp có đồng hạng do sinh viên có cùng điểm, và những lớp có ít sinh viên thì số thứ hạng cũng ít hơn.

c, điểm số theo mã môn học

In [20]:

query = '''
SELECT student_id, name, class, course_id, score,
       RANK() OVER (PARTITION BY course_id ORDER BY score DESC) AS course_rank
FROM student
ORDER BY course_id, course_rank;
'''

df_course_rank = pd.read_sql_query(query, conn)
print(df_course_rank)


   student_id               name     class  course_id  score  course_rank
0           1  Nguyen Minh Hoang  May Tinh         12    6.7            1
1           3       Pham Van Nam  Toan Tin         26    7.9            1
2           9     Duong Huu Phuc   Kinh Te         26    7.2            2
3          10       Cao Thi Hanh  May Tinh         26    7.0            3
4           2       Tran Thi Lan   Kinh Te         34    9.2            1
5           7      Bui Tien Dung   Kinh Te         34    9.2            1


1. Môn course_id = 12( giải tích)

Nguyễn Minh Hoàng: 6.7 → hạng 1
→ Đây là sinh viên duy nhất trong môn này, nên được xếp hạng 1.

2. Môn course_id = 26( tin học)

Phạm Văn Nam: 7.9 → hạng 1 (điểm cao nhất)
Dương Hữu Phúc: 7.2 → hạng 2
Cao Thị Hạnh: 7.0 → hạng 3

→ Môn này có 3 sinh viên, thứ hạng giảm dần theo điểm số.

3. Môn course_id = 34( thống kê)

Trần Thị Lan: 9.2 → hạng 1
Bùi Tiến Dũng: 9.2 → hạng 1

→ Hai sinh viên đồng điểm cao nhất, nên đồng hạng 1.

top 3 cao nhất, thấp nhất theo điểm số

In [21]:
query = '''
WITH ranked_students AS (
    SELECT student_id, name, class, course_id, score,
           RANK() OVER (ORDER BY score DESC) AS rank_desc,
           RANK() OVER (ORDER BY score ASC) AS rank_asc
    FROM student
)
SELECT * FROM ranked_students
WHERE rank_desc <= 3 OR rank_asc <= 3
ORDER BY rank_desc;
'''
df_top_all = pd.read_sql_query(query, conn)
print(" Top 3 cao nhất & thấp nhất theo điểm số toàn bộ:")
print(df_top_all)
print("\n")

 Top 3 cao nhất & thấp nhất theo điểm số toàn bộ:
   student_id               name     class  course_id  score  rank_desc  \
0           2       Tran Thi Lan   Kinh Te         34    9.2          1   
1           7      Bui Tien Dung   Kinh Te         34    9.2          1   
2           3       Pham Van Nam  Toan Tin         26    7.9          3   
3           9     Duong Huu Phuc   Kinh Te         26    7.2          4   
4          10       Cao Thi Hanh  May Tinh         26    7.0          5   
5           1  Nguyen Minh Hoang  May Tinh         12    6.7          6   

   rank_asc  
0         5  
1         5  
2         4  
3         3  
4         2  
5         1  






Hai sinh viên Tran Thi Lan và Bui Tien Dung có điểm cao nhất là 9.2.
Do hai sinh viên có cùng điểm nên cả hai đều được xếp hạng 1 .

 Các mức điểm tiếp theo

Pham Van Nam đạt 7.9, được xếp hạng 3.

Duong Huu Phuc đạt 7.2, xếp hạng 4.

Cao Thi Hanh đạt 7.0, xếp hạng 5.



Nguyen Minh Hoang có điểm 6.7, là thấp nhất trong bảng, tương ứng hạng 6 theo thứ tự giảm dần và hạng 1 theo thứ tự tăng dần (rank_asc = 1).



Điểm số của các sinh viên dao động từ 6.7 đến 9.2.

Có hiện tượng trùng điểm cao nhất (9.2) nên hàm RANK() gán cùng thứ hạng cho hai sinh viên.

Bảng kết quả đã xác định được nhóm 3 sinh viên có điểm cao nhất và nhóm sinh viên có điểm thấp nhất trong toàn bộ dữ liệu.

top 3 cao nhất , thấp nhất theo điểm số từng lớp

In [22]:
query='''
WITH ranked_students AS (
    SELECT student_id, name, class, course_id, score,
           RANK() OVER (PARTITION BY class ORDER BY score DESC) AS class_rank_desc,
           RANK() OVER (PARTITION BY class ORDER BY score ASC) AS class_rank_asc
    FROM student
)
SELECT * FROM ranked_students
WHERE class_rank_desc <= 3 OR class_rank_asc <= 3
ORDER BY class, class_rank_desc;
'''

df_top_class = pd.read_sql_query(query, conn)
print(" Top 3 cao nhất & thấp nhất theo điểm số từng lớp:")
print(df_top_class)
print("\n")

 Top 3 cao nhất & thấp nhất theo điểm số từng lớp:
   student_id               name     class  course_id  score  class_rank_desc  \
0           2       Tran Thi Lan   Kinh Te         34    9.2                1   
1           7      Bui Tien Dung   Kinh Te         34    9.2                1   
2           9     Duong Huu Phuc   Kinh Te         26    7.2                3   
3          10       Cao Thi Hanh  May Tinh         26    7.0                1   
4           1  Nguyen Minh Hoang  May Tinh         12    6.7                2   
5           3       Pham Van Nam  Toan Tin         26    7.9                1   

   class_rank_asc  
0               2  
1               2  
2               1  
3               2  
4               1  
5               1  




Lớp Kinh Tế

Trần Thị Lan và Bùi Tiến Dũng đều đạt 9.2 điểm, cùng hạng 1 trong nhóm điểm cao nhất.
Dương Hữu Phúc đạt 7.2 điểm, nằm trong nhóm top điểm thấp hơn của lớp.

Lớp Máy Tính

Cao Thị Hạnh đạt 7.0 điểm, thuộc nhóm điểm cao của lớp.
Nguyễn Minh Hoàng đạt 6.7 điểm, nằm trong nhóm điểm thấp hơn.

Lớp Toán Tin

Phạm Văn Nam đạt 7.9 điểm, đứng hạng 1, là sinh viên có điểm cao nhất trong lớp.



* Top 3 cao nhất thấp nhất theo từng mã môn học

In [23]:
query = '''
WITH ranked_students AS (
    SELECT student_id, name, class, course_id, score,
           RANK() OVER (PARTITION BY course_id ORDER BY score DESC) AS course_rank_desc,
           RANK() OVER (PARTITION BY course_id ORDER BY score ASC) AS course_rank_asc
    FROM student
)
SELECT * FROM ranked_students
WHERE course_rank_desc <= 3 OR course_rank_asc <= 3
ORDER BY course_id, course_rank_desc;
'''

df_top_course = pd.read_sql_query(query, conn)
print(" Top 3 cao nhất & thấp nhất theo điểm số từng môn học:")
print(df_top_course)
print("\n")


 Top 3 cao nhất & thấp nhất theo điểm số từng môn học:
   student_id               name     class  course_id  score  \
0           1  Nguyen Minh Hoang  May Tinh         12    6.7   
1           3       Pham Van Nam  Toan Tin         26    7.9   
2           9     Duong Huu Phuc   Kinh Te         26    7.2   
3          10       Cao Thi Hanh  May Tinh         26    7.0   
4           2       Tran Thi Lan   Kinh Te         34    9.2   
5           7      Bui Tien Dung   Kinh Te         34    9.2   

   course_rank_desc  course_rank_asc  
0                 1                1  
1                 1                3  
2                 2                2  
3                 3                1  
4                 1                1  
5                 1                1  




 Môn có course_id = 12 (giải tích)

Nguyễn Minh Hoàng – điểm 6.7
→ Đây là sinh viên duy nhất học môn này, nên vừa là điểm cao nhất vừa là điểm thấp nhất của môn.


 Môn có course_id = 26( tin học)

Phạm Văn Nam – điểm 7.9 → cao nhất
Dương Hữu Phúc – điểm 7.2 → trung bình
Cao Thị Hạnh – điểm 7.0 → thấp nhất

→ Điểm của môn này dao động từ 7.0 đến 7.9, chênh lệch không lớn.

Môn có course_id = 34( thống kê)

Trần Thị Lan – điểm 9.2
Bùi Tiến Dũng – điểm 9.2

→ Hai sinh viên có điểm cao nhất bằng nhau, cùng đứng hạng 1 trong môn

BÀI 4: bổ sung thêm một trường graduation_date có kiểu dữ liệu là DATETIME vào bảng
student để xác định thời gian tốt nghiệp của từng bạn, trong đó thời gian tốt nghiệp được
xác định bởi thời gian hiện tại cộng với số hạng tương ứng của bạn đó tính theo điểm số.

** thêm cột graduation_date vào bảng student

In [24]:
cursor.execute('''
ALTER TABLE student ADD COLUMN graduation_date DATETIME;
''')
conn.commit()

cập nhật graduation_date dựa trên thứ hạng

In [25]:
cursor.execute('''
WITH ranked_students AS (
    SELECT student_id, RANK() OVER (ORDER BY score DESC) AS rank
    FROM student
)
UPDATE student
SET graduation_date = DATETIME('now', '+' || (
    SELECT rank FROM ranked_students WHERE ranked_students.student_id = student.student_id
) || ' months');
''')
conn.commit()

In [26]:
query = '''
SELECT student_id, name, class, course_id, score, graduation_date
FROM student
ORDER BY score DESC;
'''

df_updated = pd.read_sql_query(query, conn)
print(" Bảng `student` sau khi thêm trường `graduation_date`:")
print(df_updated)

 Bảng `student` sau khi thêm trường `graduation_date`:
   student_id               name     class  course_id  score  \
0           2       Tran Thi Lan   Kinh Te         34    9.2   
1           7      Bui Tien Dung   Kinh Te         34    9.2   
2           3       Pham Van Nam  Toan Tin         26    7.9   
3           9     Duong Huu Phuc   Kinh Te         26    7.2   
4          10       Cao Thi Hanh  May Tinh         26    7.0   
5           1  Nguyen Minh Hoang  May Tinh         12    6.7   

       graduation_date  
0  2026-05-02 07:30:14  
1  2026-05-02 07:30:14  
2  2026-07-02 07:30:14  
3  2026-08-02 07:30:14  
4  2026-09-02 07:30:14  
5  2026-10-02 07:30:14  


Bảng student đã được bổ sung thêm cột graduation_date để lưu thời gian tốt nghiệp dự kiến của sinh viên.

Dữ liệu truy vấn cho thấy mỗi sinh viên đều có một giá trị ngày giờ cụ thể, ví dụ: 2026-05-02 07:30:14( Trần Thị Lan), 2026-05-02 07:30:14.( Bùi Tiến Dùng),...

Các sinh viên có thời gian tốt nghiệp khác nhau, cho thấy dữ liệu đã được cập nhật riêng cho từng bản ghi.

Việc sắp xếp ORDER BY score DESC giúp bảng vẫn hiển thị sinh viên theo điểm số từ cao xuống thấp, đồng thời vẫn giữ được thông tin về ngày tốt nghiệp.